tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AWS/Lambda/HandlerWrapperSetHandlerIntegration.cs

```diff
@@ -0,0 +1,139 @@
+[Browsable(false)]
+[EditorBrowsable(EditorBrowsableState.Never)]
+public class HandlerWrapperSetHandlerIntegration
+{
+    private const string IntegrationName = nameof(IntegrationId.AwsLambda);
+    private static readonly ILambdaExtensionRequest RequestBuilder = new LambdaRequestBuilder();
+    private static readonly Async1Callbacks _callbacks = new Async1Callbacks();
+
+    /// <summary>
+    /// OnMethodBegin callback. The input Delegate handler is the customer's handler.
+    /// And it's already wrapped by the Amazon.Lambda.RuntimeSupport.HandlerWrapper.
+    /// Here we will try further wrap it with our DelegateWrapper class
+    /// in order to make it notify the Datadog-Extension before and after the handler's each invocation.
+    /// </summary>
+    /// <typeparam name="TTarget">Type of the target</typeparam>
+    /// <param name="instance">Instance value, aka `this` of the instrumented method</param>
+    /// <param name="handler">Instance of Amazon.Lambda.RuntimeSupport.LambdaBootstrapHandler</param>
+    /// <returns>Calltarget state value</returns>
+    internal static CallTargetState OnMethodBegin<TTarget>(TTarget instance, ref Delegate handler)
+    {
+        handler = handler.Instrument(_callbacks);
+        return CallTargetState.GetDefault();
+    }
+
+    private static string ConvertPayloadStream(Stream payloadStream)
+    {
+        var reader = new StreamReader(payloadStream, Encoding.UTF8, leaveOpen: true);
+        var result = reader.ReadToEnd();
+        payloadStream.Seek(0, SeekOrigin.Begin); // Reset the offset so that it can be read by the originally intended consumer
+        return result;
+    }
+
+    private readonly struct Async1Callbacks : IBegin1Callbacks, IReturnCallback, IReturnAsyncCallback
+    {
+        public bool PreserveAsyncContext => false;
+
+        public object OnDelegateBegin<TArg1>(object sender, ref TArg1 arg)
+        {
+            Serverless.Debug("DelegateWrapper Running OnDelegateBegin");
+            try
+            {
+                var proxyInstance = arg.DuckCast<IInvocationRequest>();
+                if (proxyInstance != null)
+                {
+                    var json = ConvertPayloadStream(proxyInstance.InputStream);
+                    var scope = LambdaCommon.SendStartInvocation(new LambdaRequestBuilder(), json, proxyInstance.LambdaContext?.ClientContext?.Custom);
+                    return new CallTargetState(scope);
+                }
+                else
+                {
+                    Serverless.Debug("DuckCast.IInvocationRequest got null proxyInstance");
+                    var scope = LambdaCommon.SendStartInvocation(new LambdaRequestBuilder(), string.Empty, null);
+                    return new CallTargetState(scope);
+                }
+            }
+            catch (Exception ex)
+            {
+                Serverless.Error("Could not send payload to the extension", ex);
+                Console.WriteLine(ex.StackTrace);
```

Comment:
Shouldn't we use a Serverless or tracer logger instead of Console?
```suggestion
                Console.WriteLine(ex.StackTrace);
```

Reply:
updated